File: zig/test/walk_bench.zig
// Benchmark for walk() on wide and deep trees. Mirrors ts/test/walk-bench.test.ts.
// Not run by default. To enable, set WALK_BENCH=1 in the environment.
//
// Run with:   WALK_BENCH=1 zig build bench
// Or:         WALK_BENCH=1 zig build test-bench

const std = @import("std");
const voxgig_struct = @import("voxgig-struct");

const Allocator = std.mem.Allocator;
const JsonValue = voxgig_struct.JsonValue;

// Build a balanced tree of maps with the given width and depth.
// Total nodes: (width^(depth+1) - 1) / (width - 1).
fn buildTree(allocator: Allocator, width: usize, depth: usize) anyerror!JsonValue {
    if (depth == 0) {
        return JsonValue{ .integer = 0 };
    }
    const out = try JsonValue.makeMap(allocator);
    var i: usize = 0;
    while (i < width) : (i += 1) {
        const key = try std.fmt.allocPrint(allocator, "k{d}", .{i});
        const child = try buildTree(allocator, width, depth - 1);
        try out.object.put(key, child);
    }
    return out;
}

fn countNodes(val: JsonValue) u64 {
    if (val != .object and val != .array) return 1;
    var n: u64 = 1;
    if (val == .object) {
        var it = val.object.data.iterator();
        while (it.next()) |kv| n += countNodes(kv.value_ptr.*);
    } else if (val == .array) {
        for (val.array.data.items) |item| n += countNodes(item);
    }
    return n;
}

// Minimal consumer: sum path length at each visit.
var bench_sink: u64 = 0;

fn benchCb(_: Allocator, _: ?[]const u8, val: JsonValue, _: JsonValue, path: []const []const u8) anyerror!JsonValue {
    bench_sink +%= path.len;
    return val;
}

fn u64LessThan(_: void, a: u64, b: u64) bool {
    return a < b;
}

fn measure(allocator: Allocator, label: []const u8, tree: JsonValue, runs: usize) !void {
    // Warm-up.
    var w: usize = 0;
    while (w < 2) : (w += 1) {
        _ = try voxgig_struct.walk(allocator, tree, benchCb, null, voxgig_struct.MAXDEPTH);
    }

    var times = try std.ArrayList(u64).initCapacity(allocator, runs);
    defer times.deinit();

    var r: usize = 0;
    while (r < runs) : (r += 1) {
        var timer = try std.time.Timer.start();
        _ = try voxgig_struct.walk(allocator, tree, benchCb, null, voxgig_struct.MAXDEPTH);
        const elapsed_ns = timer.read();
        try times.append(elapsed_ns);
    }

    std.mem.sort(u64, times.items, {}, u64LessThan);
    const median_ns = times.items[times.items.len / 2];
    const min_ns = times.items[0];
    const max_ns = times.items[times.items.len - 1];
    var sum: u128 = 0;
    for (times.items) |t| sum += t;
    const mean_ns: u64 = @intCast(sum / times.items.len);

    const nodes = countNodes(tree);
    const ns_per_node_x10: u64 = if (nodes > 0) (median_ns * 10) / nodes else 0;

    std.debug.print(
        "[walk-bench] {s}: nodes={d} runs={d} min={d:.2}ms median={d:.2}ms mean={d:.2}ms max={d:.2}ms ns/node={d}.{d} sink={d}\n",
        .{
            label,
            nodes,
            runs,
            @as(f64, @floatFromInt(min_ns)) / 1e6,
            @as(f64, @floatFromInt(median_ns)) / 1e6,
            @as(f64, @floatFromInt(mean_ns)) / 1e6,
            @as(f64, @floatFromInt(max_ns)) / 1e6,
            ns_per_node_x10 / 10,
            ns_per_node_x10 % 10,
            bench_sink,
        },
    );
}

fn benchEnabled() bool {
    const val = std.process.getEnvVarOwned(std.heap.page_allocator, "WALK_BENCH") catch return false;
    defer std.heap.page_allocator.free(val);
    return std.mem.eql(u8, val, "1");
}

test "walk-bench-wide-and-deep" {
    if (!benchEnabled()) return error.SkipZigTest;

    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const tree = try buildTree(allocator, 8, 6);
    try measure(allocator, "wide+deep (w=8,d=6)", tree, 7);
}

test "walk-bench-very-wide" {
    if (!benchEnabled()) return error.SkipZigTest;

    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const tree = try buildTree(allocator, 1000, 2);
    try measure(allocator, "wide (w=1000,d=2)", tree, 7);
}

test "walk-bench-very-deep" {
    if (!benchEnabled()) return error.SkipZigTest;

    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const tree = try buildTree(allocator, 2, 20);
    try measure(allocator, "deep (w=2,d=20)", tree, 5);
}

File: rb/walk_bench.rb
# Benchmark for walk() on a wide and deep tree.
# Not run by default. To enable:
#   WALK_BENCH=1 ruby walk_bench.rb
#
# Mirrors ts/test/walk-bench.test.ts.

require_relative 'voxgig_struct'

BENCH = ENV['WALK_BENCH'] == '1'

# Build a balanced tree of maps with given width and depth.
# Total nodes: (width^(depth+1) - 1) / (width - 1).
def build_tree(width, depth)
  return 0 if depth == 0
  out = {}
  width.times do |i|
    out["k#{i}"] = build_tree(width, depth - 1)
  end
  out
end

def count_nodes(val)
  if val.is_a?(Hash)
    n = 1
    val.each_value { |v| n += count_nodes(v) }
    n
  elsif val.is_a?(Array)
    n = 1
    val.each { |v| n += count_nodes(v) }
    n
  else
    1
  end
end

def measure(label, tree, runs)
  # Touch path to simulate a minimal consumer. Using path.length keeps the
  # work O(1) so we measure walk overhead rather than callback overhead.
  sink = 0
  cb = lambda do |_k, v, _p, path|
    sink += path.length
    v
  end

  # Warm-up.
  2.times { VoxgigStruct.walk(tree, cb) }

  times = []
  runs.times do
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    VoxgigStruct.walk(tree, cb)
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    times << (t1 - t0) / 1_000_000.0  # ms
  end

  times.sort!
  median = times[times.length / 2]
  min = times[0]
  max = times[-1]
  mean = times.inject(0.0) { |a, b| a + b } / times.length

  nodes = count_nodes(tree)
  ns_per_node = (median * 1_000_000.0) / nodes

  printf(
    "[walk-bench] %s: nodes=%d runs=%d min=%.2fms median=%.2fms mean=%.2fms max=%.2fms ns/node=%.1f sink=%d\n",
    label, nodes, runs, min, median, mean, max, ns_per_node, sink
  )
  $stdout.flush
end

if BENCH
  # ~299k nodes: width=8, depth=6.
  wide_deep = build_tree(8, 6)
  measure('wide+deep (w=8,d=6)', wide_deep, 7)

  # width=1000, depth=2 -> 1,001,001 nodes, shallow.
  wide = build_tree(1000, 2)
  measure('wide (w=1000,d=2)', wide, 7)

  # width=2, depth=20 -> 2,097,151 nodes. MAXDEPTH is 32 by default.
  deep = build_tree(2, 20)
  measure('deep (w=2,d=20)', deep, 5)
else
  puts "walk_bench skipped (set WALK_BENCH=1 to run)"
end
